### PR TITLE
subscriber: update sharded-slab to 0.1, pool hashmap allocations

### LINKED ISSUE #1005: tracing-subscriber: big RAM usage
```
fn main() {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::from_default_env().add_directive("scratchpad=info".parse().unwrap()),
        )
        .init();

    std::thread::park();
}
```

This short program consumes 9 MB of residential memory on my Mac. It really shouldn't.

### PR DESCRIPTION
**Pull request by hawkw**
## Motivation

hawkw/sharded-slab#45 changes `sharded-slab` so that the per-shard
metadata is allocated only when a new shard is created, rather than all
up front when the slab is created. This fixes the very large amount of
memory allocated by simply creating a new `Registry` without actually
collecting any traces.

## Solution

This branch updates `tracing-subscriber` to depend on `sharded-slab`
0.1.0, which includes the upstream fix.

In addition, this branch the registry from using `sharded_slab::Slab` to
`sharded_slab::Pool`. This allows us to clear hashmap allocations for
extensions in-place, retaining the already allocated maps. This should
improve `new_span` performance a bit.

Fixes #1005

### CODE REVIEW

**Comment by hawkw** [comment]
i'll also backport this to v0.1.x once it merges.

File: tracing-subscriber/src/registry/sharded.rs [hunk 1]
```diff
@@ -366,6 +342,70 @@ impl<'a> SpanData<'a> for Data<'a> {
...
+        }
+
+        static NULL_METADATA: Metadata<'static> = tracing_core::metadata! {
+            name: "",
+            target: "",
+            level: tracing_core::Level::TRACE,
+            fields: &[],
+            callsite: &NULL_CALLSITE,
+            kind: tracing_core::metadata::Kind::SPAN,
+        };
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs:368]: wharrgarbl

File: tracing-subscriber/src/registry/extensions.rs [hunk 2]
```diff
@@ -113,14 +113,16 @@ impl<'a> ExtensionsMut<'a> {
 pub(crate) struct ExtensionsInner {
     // If extensions are never used, no need to carry around an empty HashMap.
     // That's 3 words. Instead, this is only 1 word.
```
Comment (davidbarsky) [inline tracing-subscriber/src/registry/extensions.rs:115]: I think this comment is removable now due to `sharded-slab`'s Pool?

File: tracing-subscriber/src/registry/sharded.rs [hunk 3]
```diff
@@ -366,6 +342,70 @@ impl<'a> SpanData<'a> for Data<'a> {
...
+        static NULL_CALLSITE: NullCallsite = NullCallsite;
+        impl tracing_core::callsite::Callsite for NullCallsite {
+            fn set_interest(&self, _: tracing_core::subscriber::Interest) {
+                unreachable!("you somehow managed to register the null callsite?")
+            }
+
+            fn metadata(&self) -> &Metadata<'_> {
+                unreachable!("you somehow managed to access the null callsite?")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/registry/sharded.rs:359]: sure, will do

File: tracing-subscriber/src/registry/extensions.rs (hunk 2 above)
Comment (hawkw) [reply tracing-subscriber/src/registry/extensions.rs:115]: ah yup i forgot to remove the comment

**Review by davidbarsky — APPROVED** [review]

**Pull request merged**